packages/subiquity_client/lib/src/server/launcher.dart

```diff
@@ -0,0 +1,233 @@
+Map<String, String> _pythonPath(String subiquityPath) {
+  final pythonPath = (Platform.environment['PYTHONPATH'] ?? '').split(':');
+  pythonPath.add(subiquityPath);
+  pythonPath.add(p.join(subiquityPath, 'curtin'));
+  pythonPath.add(p.join(subiquityPath, 'probert'));
+  return {'PYTHONPATH': pythonPath.join(':')};
+}
+
+/// Knows the many ways with which we may want to start Subiquity.
+/// Starting the server builds on top of [Process.start], so this launcher
+/// can be customized by implementing factory constructors that adjust the
+/// [command] and [args] (and possibly some factories to enable enhancing the
+/// command line arguments, working directory and environment with the results
+/// of asynchronous computations).
+/// The example below would allow starting subiquity inside a WSL2 distro
+/// while the GUI client is running on the Windows host.
+///
+/// ```dart
+/// /// Able to start the server inside [distroName] on WSL from the Windows host.
+/// factory SubiquityLauncher.wsl(String distroName, String command,
+///     {List<String>? args,
+///     Future<void>? defer,
+///     Future<void> Function()? onProcessStart,
+///     Future<List<String>> Function()? argumentsFactory}) {
+///   return SubiquityLauncher._(
+///     'wsl',
+///     args: ['-d', distroName, command, ...?args],
+///     deferStart: defer,
+///     argsFactory: argumentsFactory,
+///     onProcessStart: onProcessStart,
+///   );
+/// }
+/// ```
+///
+/// The following would allow starting subiquity inside an LXD container while
+/// the GUI is running on the Linux host, which could be good for testing.
+///
+///```dart
+/// /// Able to start the server inside [container] on LXD from the Linux host.
+/// factory SubiquityLauncher.lxd(String container, String command,
+///     {List<String>? args, bool runAsDaemon = false, String? configFile}) {
+///   final lxcArgs = <String>[
+///     '-n',
+///     container,
+///     runAsDaemon ? '-d' : '',
+///     if (configFile != null) ...['-f', configFile],
+///     '--',
+///     command,
+///     ...?args
+///  ];
+///  return SubiquityLauncher('lxc-execute', lxcArgs);
+///}
+///```
+///
+class SubiquityLauncher {
+  /// The program the process will run.
+  final String command;
+
+  /// The program arguments.
+  final List<String>? args;
+
+  /// Async function to generate the program arguments for more advanced usage.
+  /// If both [args] and [argsFactory] are supplied, the resulting program
+  /// arguments list will be the concatenation of the [args] and the resulting
+  /// list generated by calling [argsFactory].
+  final Future<List<String>> Function()? argsFactory;
+
+  /// Optional function that will be called to regenerate the server process's
+  /// working directory.
+  final Future<String?> Function()? workingDirectoryFactory;
+
+  /// Optional function that will be called to regenerate the server environment
+  /// variables set.
+  final Future<Map<String, String>> Function()? environmentFactory;
```

Comment:
i'm hoping we could make these ordinary properties if we call getSubiquityPath() first and pass it to the launcher

Reply:
Factories are not necessary thanks to the async main.